selfservice/flow/login/handler.go

```diff
@@ -348,8 +365,44 @@ type initializeSelfServiceLoginFlowForBrowsers struct {
 //	  400: jsonError
 //	  500: jsonError
 func (h *Handler) initBrowserFlow(w http.ResponseWriter, r *http.Request, ps httprouter.Params) {
-	a, err := h.NewLoginFlow(w, r, flow.TypeBrowser)
+	a, sess, err := h.NewLoginFlow(w, r, flow.TypeBrowser)
 	if errors.Is(err, ErrAlreadyLoggedIn) {
+		if r.URL.Query().Has("login_challenge") {
+			// Always accept the Hydra login request if user has a session with the required AAL
+			hlc, err := uuid.FromString(r.URL.Query().Get("login_challenge"))
+			if err != nil || hlc == uuid.Nil {
+				h.d.SelfServiceErrorManager().Forward(r.Context(), w, r, errors.WithStack(herodot.ErrBadRequest.WithReason("The login_challenge parameter is present but invalid or zero UUID")))
+				return
+			}
+			hydra_admin_url := h.d.Config().SelfServiceFlowHydraAdminURL(r.Context())
+			if hydra_admin_url == nil {
+				h.d.SelfServiceErrorManager().Forward(r.Context(), w, r, errors.WithStack(herodot.ErrBadRequest.WithReason("We received a Hydra Login Challenge, but SelfServiceFlowHydraAdminURL is not configured")))
+				return
+			}
+			cr, err := hydraclient.AcceptHydraLoginRequest(
+				hydra_admin_url.String(),
+				hlc,
+				sess.IdentityID.String(),
+				h.d.Config().SessionPersistentCookie(r.Context()),
+				int64(h.d.Config().SessionLifespan(r.Context())/time.Second),
+				sess.AMR,
+			)
```

Comment:
Unless Ory Hydra tells us that it is ok to skip the login request:

```
			lr, err := hydraclient.GetHydraLoginRequest(hydra_admin_url, hlc)
			if err != nil {
			    // ...
			}
			if lr.Skip {
			    // ...
			} else {
			    // ...
			}
```

we should not auto-accept the login. Because it is possible that Ory Hydra is **demanding** us to show the login screen.

See this code here: https://github.com/ory/hydra-login-consent-node/blob/1496070c9bc5ddcefa327a7311d288c0ede60dcc/src/routes/login.ts#L26-L43

An important question here too is that the reference implementation uses the subject ( https://github.com/ory/hydra-login-consent-node/blob/1496070c9bc5ddcefa327a7311d288c0ede60dcc/src/routes/login.ts#L41 ) which is coming from Ory Hydra, not the own session store.

I think that Ory Hydra has a protection in place which would prevent us from Ory Hydra expecting one subject, and Ory Kratos sending another. But we should double-check that!

Long story short:

1. You need to fetch the login request
2. If skip is true, you can auto-accept it with the following requirements:
    1. No session exists yet -> the `subject` value from Ory Hydra's login request response is used. See [example](https://github.com/ory/hydra-login-consent-node/blob/1496070c9bc5ddcefa327a7311d288c0ede60dcc/src/routes/login.ts#L41)
   2. A session exists and `kratos.identity.id == hydra.subject` -> the identity id from the Ory Kratos session is used
   3. A session exists and `kratos.identity.id != hydra.subject`:
       * Either send `kratos.identity.id` to Ory Hydra, which most likely will end up in an error or;
       * Invalidate the session and continue to the login form
3. If skip is false you need to:
    1. If a session exists, log the user out, and perform a fresh log in
   2. If a session does not exist, continue to the login form

Reply:
Thank you for your insightful review!

Regarding the login logic, this is how I implemented it in the update that I just submitted:

> 2. If skip is true, you can auto-accept it with the following requirements:
>     i. No session exists yet -> the subject value from Ory Hydra's login request response is used. See example

I don't think this is ever the case, because we handle the Hydra challenge either if AlreadyLoggedIn (session exists) or in the post-login hook (session exists; it has just been created).

> 2.i and 2.ii

I believe the implementation satisfies both scenarios by always using kratos.identity.id. I just need to add a test for the error when there's a mismatch (will do over the weekend).

> 3. If skip is false you need to:
>     i. If a session exists, log the user out, and perform a fresh log in

I implemented this by adding a `refresh=true` to the login request initialization. This should have the added benefit that it won't "sign the user out of github when they don't complete a circleci login".

Regarding registration, no changes were necessary because the user always enters a password during registration, and the login challenge is handled in the post-registration hook.

Please let me know what you think :)